librz/analysis/arch/8051/8051_il.c

```diff
@@ -0,0 +1,455 @@
+// SPDX-FileCopyrightText: 2021 Basstorm <?>
+// SPDX-FileCopyrightText: 2022 imbillow <billow.fun@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_il/rz_il_opbuilder_begin.h>
+#include "8051_op.h"
+
+static const char *i8051_registers_str[0xff] = {
+	[I8051_R0] = "r0",
+	[I8051_R1] = "r1",
+	[I8051_R2] = "r2",
+	[I8051_R3] = "r3",
+	[I8051_R4] = "r4",
+	[I8051_R5] = "r5",
+	[I8051_R6] = "r6",
+	[I8051_R7] = "r7",
+	[I8051_SP] = "sp",
+	[I8051_DPTR] = "dptr",
+	[I8051_PCON] = "pcon",
+	[I8051_TCON] = "tcon",
+	[I8051_TMOD] = "tmod",
+	[I8051_TL0] = "tl0",
+	[I8051_TL1] = "tl1",
+	[I8051_TH0] = "th0",
+	[I8051_TH1] = "th1",
+	[I8051_PC] = "pc",
+	[I8051_PSW] = "psw",
+	[I8051_Z] = "z",
+	[I8051_OV] = "ov",
+	[I8051_RS0] = "rs0",
+	[I8051_RS1] = "rs1",
+	[I8051_N] = "n",
+	[I8051_AC] = "ac",
+	[I8051_CY] = "cy",
+	[I8051_A] = "a",
+	[I8051_B] = "b",
+	NULL,
+};
+
+static const bool i8051_register_is_bit[0xff] = {
+	false,
+	[I8051_Z] = true,
+	[I8051_OV] = true,
+	[I8051_RS0] = true,
+	[I8051_RS1] = true,
+	[I8051_N] = true,
+	[I8051_AC] = true,
+	[I8051_CY] = true,
+};
+
+#define ADD3(a, b, c) ADD(a, ADD(b, c))
+#define SUB3(a, b, c) SUB(SUB(a, b), c)
+#define AND3(a, b, c) AND(a, AND(b, c))
+
+#define LO_BYTE(x) CAST(8, IL_FALSE, x)
+#define HI_BYTE(x) CAST(8, IL_FALSE, SHIFTR0(x, U16(8)))
+
+#define VARG_8051(x) VARG(i8051_registers_str[x])
+#define VAL_A        VARG_8051(I8051_A)
+#define VAL_B        VARG_8051(I8051_B)
+#define VAL_CY       VARG_8051(I8051_CY)
+#define VAL_CY8      BOOL_TO_BV(VARG_8051(I8051_CY), 8)
+
+static RzILOpPure *val_register(I8051OpAddressing *a) {
+	switch (a->d.reg) {
+	case I8051_PC:
+		return U16(a->pc);
+		//	case I8051_CY:
+		//	case I8051_AC:
+		//	case I8051_OV:
+		//		return BOOL_TO_BV(VARG(i8051_registers_str[a->d.reg]), 8);
+	default:
+		return VARG(i8051_registers_str[a->d.reg]);
+	}
+}
+
+static inline RzILOpPure *bv_i(RzILOpBitVector *v, ut8 i) {
+	return NON_ZERO(LOGAND(v, U8(1 << i)));
+}
+
+static inline RzILOpPure *bv_i_bv(RzILOpBitVector *v, ut8 i) {
+	return SHIFTR0(LOGAND(v, U8(1 << i)), U8(i));
+}
+
+static inline RzILOpEffect *set_reg(I8051Registers reg, RzILOpBitVector *v) {
+	return SETG(i8051_registers_str[reg], v);
+}
+
+static inline RzILOpEffect *set_a(RzILOpBitVector *v) {
+	return set_reg(I8051_A, v);
+}
+
+static inline RzILOpEffect *set_b(RzILOpBitVector *v) {
+	return set_reg(I8051_B, v);
+}
+
+static inline RzILOpEffect *set_cy(RzILOpBitVector *v) {
+	return set_reg(I8051_CY, v);
+}
+
+static RzILOpPure *get_any(I8051OpAddressing *a) {
+	switch (a->mode) {
+	case I8051_ADDRESSING_REGISTER:
+		return val_register(a);
+	case I8051_ADDRESSING_DIRECT:
+		return LOAD(U16(a->d.addr));
+	case I8051_ADDRESSING_INDIRECT:
+		return LOAD(CAST(16, IL_FALSE, get_any(a->d.indirect)));
```

Comment:
You could just use `UNSIGNED(16, get_any(a->d.indirect))` here. It is more concise and readable.

Reply:
Similarly for the rest of the occurrences of `CAST`.